rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+
+#### Portability Layer to Other Runners (aka Orchestrators) {#portability-layer-to-other-runner-orchestrators}
+
+This portability layer in tfx refers to various DagRunners which runs TFX
+pipelines on various external orchestrators, the _portable_ module and the
+abstraction between CLI and Dag runner. These libraries will comply with
+backward compatibility.
+
+Notice that several Runners are experimental by their nature, or slated to be
+deprecated. Specifically, we plan to support the following runners with backward
+compatibility:
+
+*   LocalDagRunner
+*   KubeflowV2DagRunner
+
+The following Dag runners will remain experimental:
+
+*   AirflowDagRunner
+    *   Honestly speaking, support of this one is on the fence. Due to lack of
+        proper configuration for a distributed backend support, it’s not much
+        more powerful than the LocalDagRunner when used out-of-box. However,
+        some partner teams managed to connect this to a distributed backend.
+        Until that is supported in TFX, we don’t think this one is stable.
+
+The following dag runners are slated to be deprecated and therefore won’t
+comply:
+
+*   BeamDagRunner: this will be replaced by the LocaDagRunner
+*   KubeflowDagRunner: this dag runner implementation based on KFP v1 API will
+    be replaced by the V2 dag runner above.
+
+#### FrontEnd {#frontend}
+
+Similar to Orchestrator, for the upcoming TFX::OSS frontend (first version as
+plugin to Tensorboard), no backward/forward compatibility is guaranteed.
+
+#### Command Line Interface (CLI) {#command-line-interface-cli}
+
+Command groups to fulfill key CUJs will be backward compatible, notably:
+
+*   _pipeline_
+*   _run_
+
+_The template_ command group will remain experimental.
+
+### ML-Metadata (MLMD) {#ml-metadata-mlmd}
+
+We are proposing that MLMD project enters the same 1.0 readiness and
+compatibility as TFX. The rest of this section will discuss various aspects of
+MLMD.
+
+#### Client Libraries {#client-libraries}
+
+Several protobuf structs from the `metadata_store.proto` are used and exposed by
+TFX to define Artifacts and need to subject to the compliance requirements.
+
+#### Protobuf & RPC Interface {#protobuf-&-rpc-interface}
+
+All protobufs w/o experimental or internal in their names/packages/module shall
+be SemVer compliant.
+
+MLMD has a list of gRPC interfaces. These interfaces shall be backward
+compatible, unless explicitly called out in protobuf comment.
+
+#### Database {#database}
+
+All distributions of MLMD run on top of a database. Right now, SQLite and MySQL
+are available for external users.
+
+All data model schemas for Sqlite and MySQL backends shall be backward
+compatible with the MLMD schema migration utilities.
+
+### TFX-BSL & TFXIO {#tfx-bsl-&-tfxio}
+
+Only code in the
+_[‘public’](https://github.com/tensorflow/tfx-bsl/tree/master/tfx_bsl/public)_
+module in TFX-BSL is considered public to external users.
+
+#### TFXIO Usage {#tfxio-usage}
```

Comment:
Would be nice to have the TFXIO API be tweakable and stable on API,

We are checking alternative data formats due to tf.examples sometimes not being easy to work with on wide-row datasets or to work with frameworks outside of Tensorflow (for example: XgBoost may not work great w Tf.example). 

Would be nice to have TFXIO be a public interface on TFX where we can inject custom TFXIO classes to the pipeline as needed (similar to DataView).

Reply:
@brills to comment on TFXIO API stability.

Reply:
CC @newtonle

Reply:
Circling back to this: do you think TFXIO stability is a blocker for other API to enter v1 stability? We can do it in a follow up version as long as it does not require we change other parts of the system, which I think it won't?

Reply:
sgtm!